core/shared/src/main/scala/fs2/internal/Algebra.scala

```diff
@@ -62,6 +65,7 @@ private[fs2] object Algebra {
             }
           ).asInstanceOf[AlgEffect[G, O, R]]
 
+        case r: RunLeg[F, x, O]   => r.asInstanceOf[AlgEffect[G, O, R]]
```

Comment:
I don't think so this is sound. The inner stream has to be translated instead of cast.